go/stellar/stellarsvc/frontend.go

```diff
@@ -239,16 +244,218 @@ func (s *Server) LinkNewWalletAccountLocal(ctx context.Context, arg stellar1.Lin
+	case stellar1.PaymentSummaryType_DIRECT:
+		return s.transformPaymentDirect(ctx, acctID, p.Direct())
+	case stellar1.PaymentSummaryType_RELAY:
+		return s.transformPaymentRelay(ctx, acctID, p.Relay())
+	default:
+		return nil, fmt.Errorf("unrecognized payment type: %s", typ)
+	}
+}
+
+func (s *Server) transformPaymentStellar(ctx context.Context, acctID stellar1.AccountID, p stellar1.PaymentSummaryStellar) (*stellar1.PaymentLocal, error) {
+	loc, err := newPaymentLocal(p.TxID, p.Ctime, p.Amount, p.From, p.To, acctID)
+	if err != nil {
+		return nil, err
+	}
+
+	loc.Source = p.From.String()
+	loc.SourceType = ParticipantTypeStellar
+	loc.Target = p.To.String()
+	loc.TargetType = ParticipantTypeStellar
+
+	loc.Status = stellar1.PaymentStatus_COMPLETED
+	loc.StatusDescription = strings.ToLower(loc.Status.String())
+
+	return loc, nil
+}
+
+func (s *Server) transformPaymentDirect(ctx context.Context, acctID stellar1.AccountID, p stellar1.PaymentSummaryDirect) (*stellar1.PaymentLocal, error) {
+	loc, err := newPaymentLocal(p.TxID, p.Ctime, p.Amount, p.FromStellar, p.ToStellar, acctID)
+	if err != nil {
+		return nil, err
+	}
+
+	loc.Worth, loc.WorthCurrency, err = s.formatWorth(ctx, p.DisplayAmount, p.DisplayCurrency)
+	if err != nil {
+		return nil, err
+	}
+
+	loc.Source, loc.SourceType = s.lookupUsernameFallback(ctx, p.From.Uid, p.FromStellar)
+
+	if p.To != nil {
+		loc.Target, loc.TargetType = s.lookupUsernameFallback(ctx, p.To.Uid, p.ToStellar)
+	}
+
+	loc.Status = p.TxStatus.ToPaymentStatus()
+	loc.StatusDescription = strings.ToLower(loc.Status.String())
+	loc.StatusDetail = p.TxErrMsg
+
+	loc.Note, loc.NoteErr = s.decryptNote(ctx, p.TxID, p.NoteB64)
+
+	return loc, nil
+}
+
+func (s *Server) transformPaymentRelay(ctx context.Context, acctID stellar1.AccountID, p stellar1.PaymentSummaryRelay) (*stellar1.PaymentLocal, error) {
+	loc, err := newPaymentLocal(p.TxID, p.Ctime, p.Amount, p.FromStellar, "", acctID)
+	if err != nil {
+		return nil, err
+	}
+
+	loc.Worth, loc.WorthCurrency, err = s.formatWorth(ctx, p.DisplayAmount, p.DisplayCurrency)
+	if err != nil {
+		return nil, err
+	}
+
+	loc.Source, loc.SourceType = s.lookupUsernameFallback(ctx, p.From.Uid, p.FromStellar)
+
+	if p.To != nil {
+		name, err := s.lookupUsername(ctx, p.To.Uid)
+		if err != nil {
+			s.G().Log.CDebugf(ctx, "recipient lookup failed: %s", err)
+			return nil, errors.New("recipient lookup failed")
+		}
+		loc.Target = name
+		loc.TargetType = ParticipantTypeKeybase
+	} else {
+		loc.Target = p.ToAssertion
+		loc.TargetType = ParticipantTypeKeybase
+	}
+
+	if p.TxStatus != stellar1.TransactionStatus_SUCCESS {
+		// If the funding tx is not complete
+		loc.Status = p.TxStatus.ToPaymentStatus()
+		loc.StatusDetail = p.TxErrMsg
+	} else {
+		loc.Status = stellar1.PaymentStatus_CLAIMABLE
+		loc.StatusDetail = "Waiting for the recipient to open the app to claim, or the sender to yank."
+	}
+	if p.Claim != nil {
+		loc.Status = p.Claim.TxStatus.ToPaymentStatus()
```

Comment:
This has the same problem. If someone fails a claim, it's still claimable, there's just also an error.

Reply:
Maybe there's no point trying to pack this into a `status` field. Maybe it would be clearer and easier to have something like
```
complete bool
canCancel bool
canClaim bool
claimPending bool
lastError option<string>
```

Reply:
Well, then frontend has to do work to display the status...

Reply:
I think a simplification for this whole process would be for a claimed relay payment to just be returned as a PaymentSummaryDirect.  There's no need to know that it was a relay payment.

Reply:
That's interesting. But what does it simplify? A completed relay is not a case that we are having a problem with. Also we may want to leave the option open to show both txIDs and the user's intention of the to assertion.